src/compas_ags/diagrams/forcediagram.py

```diff
@@ -301,27 +319,50 @@ def constraints_from_dual(self, tol=10e-4):
         ForceDiagram is modified in place.
         """
         edge_index = self.dual.edge_index()
+        _edges = list(self.edges())
         ordered_edges = self.ordered_edges(self.dual)
+        edges_orient = []
 
-        for edge in self.edges_where_dual({'is_ind': True}):
+        for edge in self.edges_where_dual({'is_ind': True}):  # Fix vertices of dual independent edge
             self.vertices_attribute('is_fixed', True, keys=[*edge])
-            self.edge_attribute(edge, 'has_fixed_orientation', True)
+            edges_orient.append(edge)
 
-        for edge in self.edges_where_dual({'has_fixed_orientation': True}):
-            self.edge_attribute(edge, 'has_fixed_orientation', True)
+        for edge in self.edges_where_dual({'is_load': True}):  # If loads are orthogonal the force dual edge gets constrained
+            self.edge_attribute(edge, 'is_load', True)
+            edges_orient.append(edge)
             sp, ep = self.edge_coordinates(*edge)
             if abs(sp[0] - ep[0]) < tol:
                 self.vertices_attribute('is_fixed_x', value=True, keys=[*edge])
             if abs(sp[1] - ep[1]) < tol:
                 self.vertices_attribute('is_fixed_y', value=True, keys=[*edge])
 
+        for edge in self.edges_where_dual({'is_reaction': True}):
+            self.edge_attribute(edge, 'is_reaction', True)
+            edges_orient.append(edge)
+
+        for form_edge in self.dual.edges():
+            target_vector = self.dual.edge_attribute(form_edge, 'target_vector')
+            index = edge_index[form_edge]
+            force_edge = ordered_edges[index]
+            if target_vector is not None:
+                edges_orient.append(force_edge)
+
+        for edge in edges_orient:
+            if edge in _edges:
+                pass
+            else:
+                edge = (edge[1], edge[0])
```

Comment:
why not
```python
if edge not in _edges:
    edge = (edge[1], edge[0])
```

Reply:
Yes, this is a problem because here I am fixing the orientation of these edges, and for that I get the coordinates from: `self.edge_coordinates(u, v)` and if instead of `(u, v)` we write `(v, u)` it will returns the inverse of the "natural" direction of the edge which brings problems when updating the diagram using the `parallelise` function. And this (u,v) (v,u) rotation happens when we order the edges to match the form diagram etc... So I had to include that